caikit_tgis_backend/tgis_connection.py

```diff
@@ -0,0 +1,180 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Encapsulate the creation of a TGIS Connection"""
+
+# Standard
+from dataclasses import dataclass
+from typing import Optional
+import os
+
+# Third Party
+import grpc
+
+# First Party
+from caikit.core.toolkit.errors import error_handler
+import alog
+
+# Local
+from .protobufs import generation_pb2_grpc
+
+log = alog.use_channel("TGCONN")
+error = error_handler.get(log)
+
+
+@dataclass
+class TLSFilePair:
+    cert_file: str
+    key_file: str
+
+
+@dataclass
+class TGISConnection:
+
+    # Class members
+    hostname: str
+    ca_cert_file: Optional[str] = None
+    client_tls: Optional[TLSFilePair] = None
+    _client: Optional[generation_pb2_grpc.GenerationServiceStub] = None
+
+    # Class constants
+    HOSTNAME_KEY = "hostname"
+    HOSTNAME_TEMPLATE_MODEL_ID = "model_id"
+    CA_CERT_FILE_KEY = "ca_cert_file"
+    CLIENT_CERT_FILE_KEY = "client_cert_file"
+    CLIENT_KEY_FILE_KEY = "client_key_file"
+
+    @classmethod
+    def from_config(cls, model_id: str, config: dict) -> Optional["TGISConnection"]:
+        """Create an instance from a connection template and a model_id"""
+        hostname = config.get(cls.HOSTNAME_KEY)
+        if hostname:
+            hostname = hostname.format(
+                **{
+                    cls.HOSTNAME_TEMPLATE_MODEL_ID: model_id,
```

Comment:
this should always be 1:1 mapping between model hostname and model ids, with no duplicates right?

Reply:
Well, it's certainly extremely possible to misconfigure it, but to configure it correctly, it _should_ produce a unique `hostname` for each model. That said, I intentionally did _not_ check that the template actually includes `{model_id}` so that this field can still be used to implement "use this hostname, regardless of the specific model_id." That's how it's being used currently, so this preserves the "exactly one TGIS" usecase for backwards compatibility.